File: LeetCode/1014. 最佳观光组合.py
class Solution:
    def maxScoreSightseeingPair(self, A: List[int]) -> int:
        preMax = 0
        ans = 0
        a_size = len(A)
        for i in range(a_size):
            if A[i] - i + preMax > ans:
                ans = A[i] - i + preMax        
        if A[i] + i > preMax:
            preMax = A[i] + i
        return ans

File: LeetCode/1014. 最佳观光组合.java
class Solution {
    public int maxScoreSightseeingPair(int[] A) {
        int maxValue = 0;
        for(int i = 0,j = A.length -1; i < j;) {
            if ( A[i]+A[j]-(j-i) > maxValue ) {
                maxValue = A[i]+A[j]-(j-i);
            }
            if (A[i] > A[j]) {
                j --;
            }
            else if(A[i] == A[j]) {
                if (A[i+1] >= A[j-1]) {
                    i ++;
                }
                else {
                    j --;
                }
            }
            else {
                i ++;
            }
        }

        return maxValue;
    }
}